src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -541,46 +474,30 @@ public Completable stop() {
 
     private void stopConnection(String errorMessage) {
         RuntimeException exception = null;
-        hubConnectionStateLock.lock();
+        this.state.lock();
         try {
             // errorMessage gets passed in from the transport. An already existing stopError value
             // should take precedence.
-            if (stopError != null) {
-                errorMessage = stopError;
+            if (this.state.getConnectionStateUnsynchronized().stopError != null) {
+                errorMessage = this.state.getConnectionStateUnsynchronized().stopError;
             }
             if (errorMessage != null) {
                 exception = new RuntimeException(errorMessage);
                 logger.error("HubConnection disconnected with an error {}.", errorMessage);
             }
+
+            ConnectionState connectionState = this.state.getConnectionStateUnsynchronized();
             if (connectionState != null) {
                 connectionState.cancelOutstandingInvocations(exception);
-                connectionState = null;
-            }
-
-            if (pingTimer != null) {
-                pingTimer.cancel();
-                pingTimer = null;
+                connectionState.close();
+                this.state.setConnectionState(null);
             }
 
             logger.info("HubConnection stopped.");
-            hubConnectionState = HubConnectionState.DISCONNECTED;
-            handshakeResponseSubject.onComplete();
-            redirectAccessTokenProvider = null;
-            connectionId = null;
+            this.state.changeState(HubConnectionState.CONNECTED, HubConnectionState.DISCONNECTED);
             transportEnum = TransportEnum.ALL;
```

Comment:
Is this the only place transportEnum is modified outside of the ctor? If so, couldn't we just remove this and make it final? It seems wrong to "reset" this to `TransportEnum.ALL` when the HubConnection could have been constructed with a different initial value. If someone configured the HubConnection to only use long polling, whouldn't the HubConnection still try to use WebSockets if you stop and restart it? That seems like a bug.

Reply:
I see. This is overwritten in startNegotiate. startNegotiate shouldn't do that. It should be setting state.getConnectionState().negotiatedTransportEnum or something like that.

Reply:
Good catch, I thought I had fixed this already, maybe I just dreamed it.